File: Data Structures & Algorithms/invert-a-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def dfs(self,root):
        if root is None:return
        temp=root.left
        root.left=root.right
        root.right=temp
        self.dfs(root.left)
        self.dfs(root.right)
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        # top down dfs-> passing info down to chilren
        # root.left and root.right are the states that will be updated in each step and pass down
        # my function returns-> nothing!
        # my base case is ? -> root is null
        # How do I combine left and right? -> I swap them with each other
        self.dfs(root)
        return root
        

File: Data Structures & Algorithms/binary-tree-diameter/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
int maxHeight(TreeNode* root,int &d){
    if(root==NULL) return 0;
    int l=maxHeight(root->left,d);
    int r=maxHeight(root->right,d);
    d=max(d,l+r) ;
    return max(l,r)+1;
}
    int diameterOfBinaryTree(TreeNode* root) {
        int d=0;
        maxHeight(root,d);
        return d;
    }   
};
